### PR TITLE
Additions for libre-audit & postgres

### PR DESCRIPTION
**Pull request by mvandergrift**
- Includes information about how to backup and maintain libre-audit PostgreSQL.
- Adds a new navigation section Deploy | Maintain.

### CODE REVIEW

File: content/deploy/backup/audit.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    ```
+
+
+On success, the backup creates a gzip file, `audit-postgres-backup-YYYYMMDDTHHMMSS.sql.gz`.
+
+To check that the backup succeeded, unzip the files and inspect the data.
+
+## Next Steps
+
+- Test the [Restore Audit]({{< relref "../restore/audit" >}}) procedure to ensure you can recover data in case of an emergency.
```
Comment (MattDodsonEnglish) [inline content/deploy/backup/audit.md:55]:
```suggestion
```

Comment (MattDodsonEnglish) [reply content/deploy/backup/audit.md:55]: This page exists yet, does it?

File: content/deploy/maintain/audit.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+categories: ["how-to"]
+description: How to maintain Audit PostgreSQL on your Rhize deployment
+weight: 300
+menu:
+  main:
+    parent: maintain
+    identifier:
+---
+
+This guide shows you the procedure to maintain your Audit Postgres database on your Rhize Kubernetes deployment.
```
Comment (MattDodsonEnglish) [inline content/deploy/maintain/audit.md:13]: I think we can be more specific than "Maintain": what is the actual action being documented? That's what we should title it.

File: content/deploy/maintain/audit.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+## Prerequisites
+
+Before you start, ensure you have the following:
+
+- A designated backup location, for example `~/rhize-backups/libre-audit`.
```
Comment (MattDodsonEnglish) [inline content/deploy/maintain/audit.md:19]: Would this not conflict with the backups in /deploy/backup/audit? Maybe we should give it a different title

File: content/deploy/maintain/audit.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+1. Detach target partitions from main table:
+
+    ```bash
+    kubectl exec -i audit-postgresql-0 -- psql -h localhost \
+      -d audit -U <DB_USER> \
+      -c 'alter table audit_log detach partition audit_log_p20240101;'
+    ```
+
+1. Backup partition table:
```
Comment (MattDodsonEnglish) [inline content/deploy/maintain/audit.md:40]:
```suggestion
1. Backup the partition table:
```

File: content/deploy/maintain/audit.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    ```bash
+    pg_dump -U <DB_USER> -h audit-postgres-0 -p5433 \
+      --file ./audit-p20240101.sql --table public.audit_log_p20240101 audit
+    ```
+
+   On success, the backup creates a GZIP file, `audit-p20240101.sql`.
+   To check that the backup succeeded, unzip the files and inspect the data.
+
+1. Drop the partition table to remove from database:
```
Comment (MattDodsonEnglish) [inline content/deploy/maintain/audit.md:50]:
```suggestion
1. Drop the partition table to remove from the database:
```

**Review by MattDodsonEnglish — COMMENTED** [review]
@mvandergrift 

The content and form looks good. My only comments are about structure and context.

- What is the purpose of the procedure in the maintain/audit page? Why would users want to do that?  It appears to be something about "backing up partitions", but I don't want to over-summarize something I don't understand.
- Why do you think this should be in a new "maintain" section rather than in the "backup"? This seems like a backup procedure to me.

I'm not opposed to adding a new section, but an extra item does add a little more cognitive load. So if it makes sense to group under the existing "backup" , I'd rather put it there―in a separate page or grouped with the other backup audit doc, whichever makes sense.

**Comment by mvandergrift** [comment]
> @mvandergrift
> 
> The content and form looks good. My only comments are about structure and context.
> 
>     * What is the purpose of the procedure in the maintain/audit page? Why would users want to do that?  It appears to be something about "backing up partitions", but I don't want to over-summarize something I don't understand.
> 
>     * Why do you think this should be in a new "maintain" section rather than in the "backup"? This seems like a backup procedure to me.
> 
> 
> I'm not opposed to adding a new section, but an extra item does add a little more cognitive load. So if it makes sense to group under the existing "backup" , I'd rather put it there―in a separate page or grouped with the other backup audit doc, whichever makes sense.

The partition backup is really an archive process. They're backing up and then removing data from the database. We see this a different from backing up the entire database, which they'll probably never do with a sizeable dataset like this. The backup is really there for completeness. 

We could put it in the backup section, but would need to explain that there were two types of backups - one of the entire table, and then one of the partitions of the table, and then explain that there's another maintenance step of removing the recently backed up partition. Not sure if that's more or less complex?

File: content/deploy/maintain/audit.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+{{% param pre_reqs %}}
+
+
+Also, before you start, confirm you are in the right context and namespace.
+
+{{% param k8s_cluster_ns %}}
+
+## Steps
+
+Maintaining Audit PostgreSQL involves three steps:
```
Comment (MattDodsonEnglish) [inline content/deploy/maintain/audit.md:30]:
```suggestion
To archive the PostgreSQL Audit trail, follow these steps:
```

File: content/deploy/maintain/audit.md (hunk 2 above)
Comment (MattDodsonEnglish) [inline content/deploy/maintain/audit.md:13]:
```suggestion
---
title: 'Archive the PostgreSQL Audit trail'
date: '2024-03-26T11:20:56-03:00'
categories: ["how-to"]
description: How to archive a partition of the Audit trail on your Rhize deployment
weight: 300
menu:
  main:
    parent: maintain
    identifier:
---

The [audit trial]({{< relref "/deploy/audit" >}}) can generate a high volume of data, so it is a good practice to periodically _archive_ portions of it. An archive separates a portion of the data from the database and keeps it for long-term storage. This process involves the use of PostgresQL [Table Partitions](https://www.postgresql.org/docs/current/ddl-partitioning.html).
```

File: content/deploy/maintain/audit.md (hunk 3 above)
Comment (MattDodsonEnglish) [inline content/deploy/maintain/audit.md:19]:
```suggestion
- A designated backup location, for example `~/rhize-backups/libre-audit-archive`.
```

**Review by MattDodsonEnglish — COMMENTED** [review]
OK, that makes sense. I think we can keep it in a separate section.

Can you check this suggestion and commit if it looks good?

https://github.com/libremfg/libremfg.github.io/pull/96/files#r1543050044


Two  more points:

- Does the customer need to know anything about what partitions to make? Should that be added to the pre-reqs? something like "Know the range of audit data that you want to partition".
-  Wouldn't the data here be a placeholder?
   ```diff
   - alter table audit_log detach partition audit_log_p20240101;
   + alter table audit_log detach partition audit_log_p<DATE>;
   ```
   And does PostgresQL automatically create a partition going backwards from wherever that data was? If so, that's what we can put in the pre-reqs.

File: content/deploy/backup/audit.md (hunk 1 above)
Comment (mvandergrift) [reply content/deploy/backup/audit.md]: You're correct. Working on it now. ETA this afternoon. Thanks!

**Comment by mvandergrift** [comment]
>   And does PostgresQL automatically create a partition going backwards from wherever that data was? If so, that's what we can put in the pre-reqs.

Yes, just added that functionality, so the end-user doesn't need to worry about it. As you add data,  an extension will look ahead and create partitions for 6-12 intervals ahead of what you're entering. The default is "months", but that can be overridden in the charts to days, weeks years, (minutes, seconds, etc if you're weird). Hoping we don't need to explain this to anyone as it's kind of confusing, but maybe can add a one-liner about how the defaults work?

**Comment by MattDodsonEnglish** [comment]
> As you add data, an extension will look ahead and create partitions for 6-12 intervals ahead of what you're entering. The default is "months", but that can be overridden in the charts to days, weeks years, (minutes, seconds, etc if you're weird). Hoping we don't need to explain this to anyone as it's kind of confusing, but maybe can add a one-liner about how the defaults work?

At least, I can say it's a bit confusing for me! Let's keep it high level. And if there's a time to go into the details, I'd put it in #97 .

I propose:

- At the top of the procedure, put an  notice:

  ```
  {{< notice "note" >}}

  By default, Rhize automatically creates partitions in 6 month intervals

  {{< /notice >}}

- Then, in the procedure use a placeholder:

  ```
  alter table audit_log detach partition audit_log_p<ARCHIVE_DATE>
  ```

  _Where <ARCHIVE_DATE> is the date you make the archive, e.g. 20240101. Rhize automatically archives all partitions that are older than 6 months_.
  

I feel like I misunderstood something, so please correct me. Then I'll add it. The essential info is the default period and what happens to the database when the user runs the command.

File: content/deploy/maintain/audit.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+categories: ["how-to"]
+description: How to archive a partition of the Audit trail on your Rhize deployment
+weight: 300
+menu:
+  main:
+    parent: maintain
+    identifier:
+---
+
+The [audit trial]({{< relref "/how-to/audit" >}}) can generate a high volume of data, so it is a good practice to periodically _archive_ portions of it. An archive separates a portion of the data from the database and keeps it for long-term storage. This process involves the use of PostgresQL [Table Partitions](https://www.postgresql.org/docs/current/ddl-partitioning.html).
```
Comment (tomhollingworth) [inline content/deploy/maintain/audit.md:13]: Doing this speeds up the query of existing data, whilst allowing for cost effective storage of older data.

File: content/deploy/maintain/audit.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+## Prerequisites
+
+Before you start, ensure you have the following:
+
+- A designated backup location, for example `~/rhize-backups/libre-audit`.
+- Access to the [Rhize Kubernetes Environment](/deploy/install/setup-kubernetes)
+{{% param pre_reqs %}}
+
+
```
Comment (tomhollingworth) [inline content/deploy/maintain/audit.md:23]: Two new lines seem out of place

File: content/deploy/maintain/audit.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+Also, before you start, confirm you are in the right context and namespace.
+
+{{% param k8s_cluster_ns %}}
+
+## Steps
+
+To archive the PostgreSQL Audit trail, follow these steps:
+
```
Comment (tomhollingworth) [inline content/deploy/maintain/audit.md:31]: Before detatching / dumping, provide instruction on how to list the current partitions:

1. Based on your retention period query for the names of the existing partitions:

   ```bash
   kubectl exec -i audit-postgresql-0 -- psql -h localhost \ 
     -d audit -U <DB_USER> \
     -c 'SELECT * FROM x'
   ```
   
    Record the name `<PARTITION_NAME>` of the partition you wish to detach and archive.

File: content/deploy/maintain/audit.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+1. Detach target partitions from main table:
+
+    ```bash
+    kubectl exec -i audit-postgresql-0 -- psql -h localhost \
+      -d audit -U <DB_USER> \
+      -c 'alter table audit_log detach partition audit_log_p20240101;'
```
Comment (tomhollingworth) [inline content/deploy/maintain/audit.md:37]: replace audit_log_p20240101 with `<PARTITION_NAME>`

File: content/deploy/maintain/audit.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    ```
+
+1. Backup the partition table:
+
+    ```bash
+    pg_dump -U <DB_USER> -h audit-postgres-0 -p5433 \
+      --file ./audit-p20240101.sql --table public.audit_log_p20240101 audit
```
Comment (tomhollingworth) [inline content/deploy/maintain/audit.md:44]: Replace `audit_log_p20240101` with `<PARTITION_NAME>`

**Review by tomhollingworth — CHANGES_REQUESTED** [review]
A couple of notes

**Comment by mvandergrift** [comment]
@tomhollingworth @MattDodsonEnglish I believe I've addressed all the above in my latest commit.

**Review by MattDodsonEnglish — APPROVED** [review]
Looks great. Context and procedures all seem clear to me now.

**Pull request merged**